README
TimeTracker

Track time spend on tasks

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc TimeTracker

COPYRIGHT AND LICENCE

Copyright (C) 2008 Thomas Klausner

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for App::TimeTracker

{{$NEXT}}
    - remove old corner case when an unknown project is passed explicitly (Thomas Klausner)
    - Set the project attribute from config file if possible (Zakariyya Mughal)
    - Add flag to init command to set project name (Zakariyya Mughal)
    - Output projects.json using canonical setting (Zakariyya Mughal)
    - Tell users to submit bugs via GitHub (Paul Cochrane)
    - remove german KW, simplify week-start/stop calc (Thomas Klausner)


3.009     2021-01-24 19:53:24+01:00
    - add week-overview as report-option (Thomas Mantl)
    - Apply perltidy settings to all test files (Paul Cochrane)
    - min Author bundle version (Thomas Klausner)
    - Fix Dist::Zilla build problems (Paul Cochrane)


3.008     2021-01-05 12:14:08+01:00
    - fix global report test that breaks if the test were run before 14:30 (Thomas Klausner)
    - switch to Dist::Zilla::PluginBundle::Author::DOMM


3.007	2020-12-28T20:53:03
    - set github bugtracker (Thomas Klausner)


3.006   2020-12-28T20:34:29
    - test for global report (list all) and report filtered by project (Thomas Klausner)
    - Add more details to init command documentation (Paul Cochrane)
    - Extract duplicate error messages into Constants package (Paul Cochrane)
    - Add init command hint to start command help message (Paul Cochrane)
    - Test current start command help output without project (Paul Cochrane)
    - Add an initial test for report command (Paul Cochrane)
    - add --output opt to "report" command to show different fields in the report table (Thomas Klausner)


3.005   2020-11-22T17:17:58
    - Move docs to saner placed (Paul Cochrane)
    - Update cpanm installation docs (Paul Cochrane)
    - Replace references to search.cpan.org with metacpan (Paul Cochrane)
    - Remove unused GitHub::Meta option (Paul Cochrane)
    - Tell users to use GitHub issues for bug reporting (Paul Cochrane)
    - use github-issues instead of RT via dist.ini (Thomas Klausner)
    - Add initial Travis-CI configuration (Paul Cochrane)


3.004	2020-08-10T07:34:58
    - make this/last doc example a bit clearer (Thomas Klausner)
    - Document the --from, --to, --this and --last options (Paul Cochrane)
    - Fix off-by-one rounding error in Task->rounded_minutes (Paul Cochrane)
    - Add assertions to test of `current` command (Paul Cochrane)
    - Remove leading zero from integer (Paul Cochrane)
    - Purge trailing whitespace in source files (Paul Cochrane)


3.003  2020-08-02T17:40:42
    - Sort list of available commands for consistency (Paul Cochrane)
    - Stop gathering vim `.swp` files when building dist (Paul Cochrane)
    - Link to YAPC::EU 2011 talk in docs (Paul Cochrane)
    - Describe installation via tarball and git checkout separately (Paul Cochrane)
    - Bump the copyright year (Paul Cochrane)
    - Fix various typos, general cleanup (Paul Cochrane)
    - Minor pod tidy up. (Mohammad S Anwar)


3.002   2019-11-07T12:33:55
    - fix a bug in tracker report (Jozef Kutej)
    - fix PODNAME for tracker_bash_autocomplete, which properly fixes the POD link (Håkon Hægland)


3.001   2019-11-01T15:50:11
    - fix link in POD documentation. (Håkon Hægland)


3.000   2019-06-22
    - remove RT, now a standalone dist "App-TimeTracker-RT"
    - removed Post2IRC, now a standalone dist called "App-TimeTracker-TellRemote"
    - removed Git & SyncViaGit, now a standalone dist called 'App-TimeTracker-Git'
    - removed Overtime (bad API, not documented or used AFAIK)
    - removed TextNotify (not sure it still works, cannot test without a mac)


2.028	2019-05-10T16:01:11
   - Fixed previous task no longer being available for plugins (Michael Kröll)


2.027   2019-03-24T20:24:46
    - report if not in a dir containing a project (Thomas Klausner)
    - "stop" no always and properly uses the context of the stopped project (Thomas Klausner)


2.026   2019-03-12T10:57:58
    - Parsing/Encoding of config files (pepl)
    - explicitly set TZ in some tests, as reported in RT125360 (Thomas Klausner)


2.025   2018-05-13
    - Fix spacing in lists without description (Klaus Ita)


2.024   2016-02-11T10:39:11
    - report: fix --detail option (Thomas Klausner)
    - list/worked: support --fparent (Thomas Klausner)
    - report: fix accumulation of child times in deep trees (Thomas Klausner)
    - Fixed broken links in the pod (Mohammad S Anwar)


2.023   2016-01-24T21:20:58
    - pack run into eval so we can get nicer error messages (Thomas Klausner)
    - pass current config to load_attrib (Thomas Klausner)


2.022   2016-01-16T16:18:37
    - remove some warnings (github issue #16 reported by David Provost) (Thomas Klausner)
    - Add link tags to App::TimeTracker instances in POD (Paul Cochrane)
    - Improve tracker POD slightly (Paul Cochrane)
    - Fix typos in POD (Paul Cochrane)
    - Use the word "plugin" consistently (Paul Cochrane)
    - Wrap POD text more consistently (Paul Cochrane)
    - Remove trailing whitespace (Paul Cochrane)
    - --fparent: include parent in list of project (Thomas Klausner)
    - add fparent option to report (Thomas Klausner)


2.021	2015-08-08T20:46:51
    - attribute all contributors (Thomas Klausner)
    - output the JSON in its canonical form (Yanick Champoux)


2.020   2014-05-27T15:28:40+0200
    - Use bytes for creating the token digest, otherwise we'll get into trouble (Michael Kröll)
    - Message already is utf-8 characters at that point, do not double-encode (Michael Kröll)
    - aded safe_branch_name (Thomas Klausner)
    - fixed typo (David Schmidt)
    - set current time when stopping via start & reprotoing (Thomas Klausner)


2.019   2014-03-02T14:43:45+0100
    - init: store new project in projects.json (Thomas Klausner)
    - set correct stop time when doing start (Thomas Klausner)
    - find_task_files performance optimization (Lukas Rampa)
    - in cmd_stop, generate a new dynamic class based on the stopped config,

lib/App/TimeTracker.pm
package App::TimeTracker;

# ABSTRACT: time tracking for impatient and lazy command line lovers
# VERSION

use strict;
use warnings;
use 5.010;

use App::TimeTracker::Data::Task;
use DateTime;
use Moose;
use Moose::Util::TypeConstraints;
use Path::Class qw();
use Path::Class::Iterator;
use MooseX::Storage::Format::JSONpm;
use JSON::XS;

our $HOUR_RE   = qr/(?<hour>[012]?\d)/;
our $MINUTE_RE = qr/(?<minute>[0-5]?\d)/;
our $DAY_RE    = qr/(?<day>[0123]?\d)/;
our $MONTH_RE  = qr/(?<month>[01]?\d)/;
our $YEAR_RE   = qr/(?<year>2\d{3})/;

with qw(
    MooseX::Getopt
);

subtype 'TT::DateTime' => as class_type('DateTime');
subtype 'TT::RT'       => as 'Int';
subtype 'TT::Duration' => as enum( [qw(day week month year)] );

coerce 'TT::RT' => from 'Str' => via {
    my $raw = $_;
    $raw =~ s/\D//g;
    return $raw;
};

coerce 'TT::DateTime' => from 'Str' => via {
    my $raw = $_;
    my $dt  = DateTime->now;
    $dt->set_time_zone('local');
    $dt->set( second => 0 );

    if ($raw) {
        if ( $raw =~ /^ $HOUR_RE : $MINUTE_RE $/x ) {    # "13:42"
            $dt->set( hour => $+{hour}, minute => $+{minute} );
        }
        elsif ( $raw =~ /^ $YEAR_RE [-.]? $MONTH_RE [-.]? $DAY_RE $/x )
        {                                                # "2010-02-26"
            $dt->set( year => $+{year}, month => $+{month}, day => $+{day} );
            $dt->truncate( to => 'day' );
        }
        elsif ( $raw
            =~ /^ $YEAR_RE [-.]? $MONTH_RE [-.]? $DAY_RE \s+ $HOUR_RE : $MINUTE_RE $/x
            )
        {                                                # "2010-02-26 12:34"
            $dt->set(
                year   => $+{year},
                month  => $+{month},
                day    => $+{day},
                hour   => $+{hour},
                minute => $+{minute} );
        }
        elsif ( $raw =~ /^ $DAY_RE [-.]? $MONTH_RE [-.]? $YEAR_RE $/x )
        {                                                # "26-02-2010"
            $dt->set( year => $+{year}, month => $+{month}, day => $+{day} );
            $dt->truncate( to => 'day' );
        }
        elsif ( $raw
            =~ /^ $DAY_RE [-.]? $MONTH_RE [-.]? $YEAR_RE \s $HOUR_RE : $MINUTE_RE $/x
            )
        {                                                # "26-02-2010 12:34"
            $dt->set(
                year   => $+{year},
                month  => $+{month},
                day    => $+{day},
                hour   => $+{hour},
                minute => $+{minute} );
        }
        else {
            confess "Invalid date format '$raw'";
        }
    }
    return $dt;
};

MooseX::Getopt::OptionTypeMap->add_option_type_to_map( 'TT::DateTime' => '=s',
);
MooseX::Getopt::OptionTypeMap->add_option_type_to_map( 'TT::RT' => '=i', );

no Moose::Util::TypeConstraints;

has 'home' => (
    is       => 'ro',
    isa      => 'Path::Class::Dir',
    traits   => ['NoGetopt'],
    required => 1,
);
has 'config' => (
    is       => 'ro',
    isa      => 'HashRef',
    required => 1,
    traits   => ['NoGetopt'],
);
has '_current_project' => (
    is        => 'ro',
    isa       => 'Str',
    predicate => 'has_current_project',
    traits    => ['NoGetopt'],
);

has 'tags' => (
    isa     => 'ArrayRef',
    is      => 'ro',
    traits  => ['Array'],
    default => sub { [] },
    handles => {
        insert_tag => 'unshift',
        add_tag    => 'push',
    },
    documentation => 'Tags [Multiple]',
);

has '_current_command' => (
    isa    => 'Str',
    is     => 'rw',
    traits => ['NoGetopt'],
);

has '_current_task' => (
    isa    => 'App::TimeTracker::Data::Task',
    is     => 'rw',
    traits => ['NoGetopt'],
);

has '_previous_task' => (
    isa    => 'App::TimeTracker::Data::Task',
    is     => 'rw',
    traits => ['NoGetopt'],
);

sub run {
    my $self = shift;
    my $command = 'cmd_' . ( $self->extra_argv->[0] || 'missing' );

    $self->cmd_commands()
        unless $self->can($command);
    $self->_current_command($command);
    $self->$command;
}

sub now {
    my $dt = DateTime->now();
    $dt->set_time_zone('local');
    return $dt;
}

sub beautify_seconds {
    my ( $self, $s ) = @_;
    return '0' unless $s;
    my ( $m, $h ) = ( 0, 0 );

    if ( $s >= 60 ) {
        $m = int( $s / 60 );
        $s = $s - ( $m * 60 );
    }
    if ( $m && $m >= 60 ) {
        $h = int( $m / 60 );
        $m = $m - ( $h * 60 );
    }
    return sprintf( "%02d:%02d:%02d", $h, $m, $s );
}

sub find_task_files {
    my ( $self, $args ) = @_;

    my $root = $self->home;
    my ( $cmp_from, $cmp_to );

    if ( my $from = $args->{from} ) {
        my $to = $args->{to} || $self->now;
        $to->set( hour => 23, minute => 59, second => 59 ) unless $to->hour;
        $cmp_from = $from->strftime("%Y%m%d%H%M%S");
        $cmp_to   = $to->strftime("%Y%m%d%H%M%S");

        if ( $from->year == $to->year ) {
            $root = $root->subdir( $from->year );
            if ( $from->month == $to->month ) {
                $root = $root->subdir( $from->strftime("%m") );
            }
        }
    }

    my $projects;
    if ( $args->{projects} ) {
        $projects = join( '|', map { s/-/./g; $_ } @{ $args->{projects} } );
    }

    my $children;
    if ($args->{parent}) {
        my @kids = $args->{parent};
        $self->all_childs_of($args->{parent},\@kids);
        $children = join( '|', map { s/-/./g; $_ } @kids );
    }

    my $tags;
    if ( $args->{tags} ) {
        $tags = join( '|', @{ $args->{tags} } );
    }

    my @found;
    my $iterator = Path::Class::Iterator->new( root => $root, );
    until ( !$iterator || $iterator->done ) {
        my $file = $iterator->next;

        next unless -f $file;
        my $name = $file->basename;
        next unless $name =~ /\.trc$/;

        if ($cmp_from) {
            $file =~ /(\d{8})-(\d{6})/;
            my $time = $1 . $2;
            next if $time < $cmp_from;
            next if $time > $cmp_to;
        }

        if ($projects) {
            next unless ( $name =~ m/$projects/i );
        }

        if ($children) {
            next unless ( $name =~ m/$children/i );
        }

        if ($tags) {
            my $raw_content = $file->slurp;
            next unless $raw_content =~ /$tags/i;
        }

        push( @found, $file );
    }
    return sort @found;
}

sub project_tree {
    my $self = shift;
    my $file = $self->home->file('projects.json');
    return unless -e $file && -s $file;
    my $decoder  = JSON::XS->new->utf8->pretty->relaxed;
    my $projects = $decoder->decode( scalar $file->slurp );

    my %tree;
    my $depth;
    while ( my ( $project, $location ) = each %$projects ) {
        $tree{$project} //= { parent => undef, children => {} };
        # check config file for parent
        if ( -e $location ) {
            my $this_config = $decoder->decode(
                scalar Path::Class::file($location)->slurp );
            if ( my $parent = $this_config->{parent} ) {
                $tree{$project}->{parent} = $parent;
                $tree{$parent}->{children}{$project} = 1;
                next;
            }
        }
        # check path for parent
        my @parts = Path::Class::file($location)->parent->parent->dir_list;
        foreach my $dir (@parts) {
            if ( $project ne $dir and my $parent = $projects->{$dir} ) {
                $tree{$project}->{parent} = $dir;
                $tree{$dir}->{children}{$project} = 1;
            }
        }
    }

    return \%tree;
}

sub all_childs_of {
    my ($self, $parent, $collector) = @_;

    my $tree = $self->project_tree;
    my $this = $tree->{$parent};

    my @kids = keys %{$this->{children}};

    if (@kids) {
        push(@$collector, @kids);
        foreach my $kid (@kids) {
            $self->all_childs_of($kid, $collector);
        }
    }
}

1;

__END__

=head1 SYNOPSIS

Backend for the C<tracker> command. See L<tracker> and/or C<perldoc tracker> for details.

=head1 INSTALLATION

L<App::TimeTracker> is a L<Perl|http://perl.org> application, and thus requires
a recent Perl (>= 5.10). It also reuses a lot of code from
L<CPAN|http://cpan.org>.

=head2 From CPAN

The easiest way to install the current stable version of L<App::TimeTracker> is
via L<CPAN|http://cpan.org>. There are several different CPAN clients
available:

=head3 cpanminus

  ~$ cpanm App::TimeTracker
  --> Working on App::TimeTracker
  Fetching http://search.cpan.org/CPAN/authors/id/D/DO/DOMM/App-TimeTracker-2.009.tar.gz ... OK
  Configuring App-TimeTracker-2.009 ... OK
  Building and testing App-TimeTracker-2.009 ... OK
  Successfully installed App-TimeTracker-2.009
  1 distribution installed

If you don't have C<cpanminus> installed yet, L<install it right
now|http://search.cpan.org/dist/App-cpanminus/lib/App/cpanminus.pm#INSTALLATION>:

  ~$ curl -L http://cpanmin.us | perl - --sudo App::cpanminus

=head3 CPAN.pm

CPAN.pm is available on ancient Perls, and feels a bit ancient, too.

  cpan App::TimeTracker

=head2 From a tarball

To install L<App::TimeTracker> from a tarball, do the usual CPAN module
install dance:

  ~/perl/App-TimeTracker$ perl Build.PL
  ~/perl/App-TimeTracker$ ./Build
  ~/perl/App-TimeTracker$ ./Build test
  ~/perl/App-TimeTracker$ ./Build install  # might require sudo

=head2 From a git checkout

Clone the repository if you have not already done so, and enter the
C<App-TimeTracker> directory:

  ~$ git clone git@github.com:domm/App-TimeTracker.git
  ~$ cd App-TimeTracker

C<App-TimeTracker> uses L<Dist::Zilla> to build, test and install the code,
hence this must be installed first, e.g. with C<cpanm>:

  ~/path/to/App-Tracker$ cpanm Dist::Zilla

Now install the distribution's dependencies, test and install in the usual
manner for C<Dist::Zilla> projects:

  ~/path/to/App-Tracker$ dzil listdeps --missing | cpanm
  ~/path/to/App-Tracker$ dzil test
  ~/path/to/App-Tracker$ dzil install

=head1 PLUGINS

Custom commands or adaptations to your workflow can be implemented via
an "interesting" set of L<Moose>-powered plugins. You can configure
different sets of plugins for different jobs or projects.

B<Tip:> Use C<tracker plugins> to list all installed plugins. Read more
about each plugin in C<App::TimeTracker::Command::PLUGIN-NAME>.

=head2 Note about (missing) Plugins

Up to version 2.028 a lot of plugins where included in the main distribution
C<App-TimeTracker>. To make installation easier and faster, all non-core
command plugins have been moved into distinct, standalone distributions.

The following plugins are affected:

=over

=item * App::TimeTracker::Git (which also includes SyncViaGit)

=item * App::TimeTracker::RT

=item * App::TimeTracker::TellRemote (which was called Post2IRC earlier)

=item * App::TimeTracker::Overtime has been removed, while the idea is nice, the API and implementation are not good enough.

=item * App::TimeTracker::TextNotify has been removed.

=back

=head1 SOURCE CODE

=head2 git

We use C<< git >> for version control and maintain a public repository on
L<github|http://github.com>.

You can find the latest version of L<App::TimeTracker> here:

L<https://github.com/domm/App-TimeTracker>

If you want to work on L<App::TimeTracker>, add a feature, add a plugin or fix
a bug, please feel free to L<fork|http://help.github.com/fork-a-repo/> the
repo and send us L<pull requests|http://help.github.com/send-pull-requests/>
to merge your changes.

To report a bug, please use the C<< issues >> feature from GitHub.

=head2 CPAN

L<App::TimeTracker> is distributed via L<CPAN|http://cpan.org/>, the
Comprehensive Perl Archive Network. Here are a few different views of
CPAN, offering slightly different features:

=over

=item * L<https://metacpan.org/release/App-TimeTracker/>

=item * L<http://search.cpan.org/dist/App-TimeTracker/>

=back

=head1 Viewing and reporting Bugs

Please use this URL to view and report bugs:

L<https://github.com/domm/App-TimeTracker/issues>

=head1 CONTRIBUTORS

Maros Kollar, Klaus Ita, Yanick Champoux, Lukas Rampa, David Schmidt, Michael Kröll, Thomas Sibley, Nelo Onyiah, Jozef Kutej, Roland Lammel, Ruslan Zakirov, Kartik Thakore, Tokuhiro Matsuno, Paul Cochrane, David Provost, Mohammad S Anwar, Håkon Hægland, Thomas Mantl, Zakariyya Mughal
